### ISSUE TITLE
4.13.1 Release checklist

### ISSUE THREAD

**Issue by compulim** [microsoft/BotFramework-WebChat#3960]
# Checklist

## Build

1. [x] ~Bump MockBot to Bot Framework SDK release 4.13.1~
   - https://www.npmjs.com/package/botbuilder/v/4.13.1
   - (We will not bump for patch release)
1. [x] ~Bump `botframework-directlinejs` to `0.14.1`~
   - No newer version
1. [x] Bump to `4.13.1`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.13.1`
   - [x] Run `npm version --no-git-tag-version 4.13.1`
   - [x] ~Merged into `master`, the PR number is #XXXX~
      - (Will not merge as PR for `qfe`)
   - Commit is [`e9a95f8`](https://github.com/microsoft/BotFramework-WebChat/commit/e9a95f8c8bb66627894bf95db3ce6696b2718408)
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run daily pipeline manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`254023`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=254023&view=results) and commit is [`e9a95f8`](https://github.com/microsoft/BotFramework-WebChat/commit/e9a95f8c8bb66627894bf95db3ce6696b2718408)
1. [x] Wait for [`WebChat-release-testing` pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_release?_a=releases&view=mine&definitionId=17) to complete
   - Pipeline name is `Push-Release-Testing-to-GitHub-Pages`
   - The release ID is [`213`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2727)
1. [x] Add manual tests to WebChat-release-testing as needed

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [ ] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 91.0.4472.106
   - [ ] Edge 91.0.864.54
   - [ ] Firefox 89.0.1
   - [ ] IE11 (Internet Explorer mode in Windows 10 21390.2025)
   - [ ] macOS Safari ???
   - [ ] iOS Safari 14.5.1
   - [ ] Android Chrome 91.0.4472.101
1. [x] Test specific fixes related to `4.13.1` and previous releases
   
   - Strip Markdown [#3917](https://github.com/microsoft/BotFramework-WebChat/pull/3917)

## Release
1. [ ] Verify on WebChat-release-testing
1. [ ] Make sure you are on `master` ~or `qfe`~ branch, run `git status` to check
1. [ ] `git pull`
1. [ ] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.13.1`
1. [ ] `git log`
   - Verify the latest commit is `585da0a`
1. [ ] `git tag v4.13.1`
1. [ ] `git push -u upstream v4.13.1`
   - You do not need to kick off a build again, use the previous build
1. [ ] Create a new GitHub release, copy entries from `CHANGELOG.md`
   - [ ] Subresource Integrity can be generated by
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.13.1/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
      - From local: `cat webchat.js | openssl dgst -sha384 -binary | openssl base64 -A`
   - [ ] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -L -o botframework-directlinespeech-sdk-4.13.1.tgz https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.13.1.tgz
         curl -L -o botframework-webchat-4.13.1.tgz https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.13.1.tgz
         curl -L -o botframework-webchat-core-4.13.1.tgz https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.13.1.tgz
         curl -L -o botframework-webchat-api-4.13.1.tgz https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.13.1.tgz
         curl -L -o botframework-webchat-component-4.13.1.tgz https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.13.1.tgz
         ```
1. [ ] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [XXXXXX](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=XXXXXX&view=results) release number is `32` and commit is `xxxxxxx`
   - Retain the build indefinitely
1. [ ] Kick off release to CDN (CDN CUTOFF TIME IS 2PM PST)
   1. [ ] Prepare the email for approval
   - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
   - The build number is [`XXXXXX`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=XXXXXXview=results), release number is [`xxxx`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=xxxx) and commit is `xxxxxxx`
   - Script build number is `131156` (this is fixed)
   1. [ ] Send reminder email to approvers
   - Retain the build indefinitely

## Post-release verification - complete within 30mins / 1hr of the release

- [ ] Test using `webchat-release-testing`
   1. [ ] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [ ] `01.create-react-app`
      1. [ ] Nuke `01.create-react-app/node_modules`
      1. [ ] `npm install`
      1. [ ] `npm install botframework-webchat@4.13.1` (just install the bundle package)
      1. [ ] `npm run build`
   1. [ ] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.13.1, with subresource integrity
         ```
         <script
            crossorigin="anonymous"
            integrity="sha384-"
            src="https://cdn.botframework.com/botframework-webchat/latest/webchat-es5.js"
         ></script>
         ```
   1. [ ] `npx serve` (at repo root)
   1. [ ] Go to http://localhost:5000/ to test, including IE11

## Notification to interested parties

- [ ] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/381
- [ ] Notify related parties for the following fixes
   - [ ] SDK team
   - [ ] Omnichannel
   - [ ] Pooja
- [ ] Update root README.md with feature notes -- Note: PR will be combined with post-release checklist PR
- [ ] Update https://github.com/microsoft/BotFramework-sdk/issues/6245
   
---

# Post-release checklist

These are chores that we should do before starting the cycle to reduce ripple effects if we do it in mid-cycle.

Tips: 

- Clean your repo before start
- Remove `node_modules` from all folder
   - `git clean -fdx`
- Never delete `package-lock.json`
- If you mess it up, tableflip and redo
- In `component/package.json`
   - Remove reference to `botframework-webchat-core` by hand-modifying `package.json`
   - Then, `npm install` (symlinks will be broken afterward)
   - Then, add those references back by hand-modifying `package.json`
   - This also applies for other packages with similar dependencies/symlinks
   - To build afterward, do tableflip to rebuild those symlinks

## Applies to all releases

> This list should be copied to versions in the future.

- [ ] ~If on QFE branch, make sure `CHANGELOG.md` and version number bump is cherry-picked to `master`~
   - [ ] ~`git checkout master`~
   - [ ] ~`git cherry-pick XXX` (the commitish for bumping version number and `CHANGELOG.md`)~
- [ ] Correct the date for 4.13.1 in `CHANGELOG.md`
   - There could be last minute fixes that could push the planned date later than the one in `CHANGELOG.md`
- [ ] Bump `package.json` to `4.13.1-0`
   - Run `npm version prepatch --no-git-tag-version`
- [ ] Update [`servicingPlan.json`](https://github.com/microsoft/BotFramework-WebChat/blob/master/packages/embed/servicingPlan.json)
   - Add deprecation notes for previous versions
   - Subresource integrity hash from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.13.1
- [ ] Update all samples to use `4.13.1`
   - Some samples are pointing to GitHub Releases because the sample need new features from daily build
   - Search "https://github.com/microsoft/BotFramework-WebChat/releases/download/"
      - And replace with "https://cdn.botframework.com/botframework-webchat/latest/"

The PR is https://github.com/microsoft/BotFramework-WebChat/pull/XXX

### Bump all dependencies to latest version

On each `package.json` under root and `/packages/`:

- [ ] All NPM packages (`@babel/*`, `jest`, `lerna`, `typescript`, `webpack`, etc) in PR #XXX
   - Don't bump the following packages
      - `react`, `react-dom`, `redux`, `redux-saga`, `react-film`, `react-scroll-to-bottom`, `react-say`, `react-dictate-button`, `@types/react`
      - `source-map-loader` (because of webpack)
      - `serve`
      - `rxjs` on `playground`
      - `webpack@4`, [`copy-webpack-plugin@6`](https://npmjs.com/package/copy-webpack-plugin), [`html-webpack-plugin@4`](https://npmjs.com/package/html-webpack-plugin)
         - Will update in #3914
         - Starting from version 5, they dropped Node.js polyfills for browsers
      - `adaptivecards`, `microsoft-cognitiveservices-speech-sdk`
         - We will bump them in separate pull requests
      - `expect@25` in `/packages/test/*`
         - The newer version requires Node.js (`graceful-fs`)
      - `sanitize-html@1`
         - Newer version is not ES5 compatible
- [ ] Run `npm audit fix` to make sure everything is fixed
- [ ] Test under IE11 to make sure all dependencies are working

### Bump Docker image

> The Docker image can be found at root `docker-compose.yml` and `Dockerfile*`.

- [x] ~Docker container for headless Chrome (#XXX)~
   - Not bumping after QFE
   - They recently moved from `3.14.159-xxx` tag scheme to a more sensible `87.0` tag scheme
   - Tags can be found at https://hub.docker.com/r/selenium/node-chrome/tags
   - Preferably in separate PR because screenshots change can be large occasionally
   - Run tests locally, as the screenshots can be slightly different
- Note on 11/10/2020: `4.0.0-alpha-7-prerelease-20200907` is still prerelease; therefore not bumping (corinagum)
